src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContains.cs

```diff
@@ -0,0 +1,347 @@
+        private const string Add = nameof(Add);
+        private const string Remove = nameof(Remove);
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(DoNotGuardSetAddOrRemoveByContainsTitle)),
+            CreateLocalizableResourceString(nameof(DoNotGuardSetAddOrRemoveByContainsMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(DoNotGuardSetAddOrRemoveByContainsDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out var symbols))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(OnConditional, OperationKind.Conditional);
+
+            void OnConditional(OperationAnalysisContext context)
+            {
+                var conditional = (IConditionalOperation)context.Operation;
+
+                if (!symbols.HasApplicableContainsMethod(conditional.Condition, out var containsInvocation, out bool containsNegated) ||
+                    !symbols.HasApplicableAddOrRemoveMethod(conditional.WhenTrue.Children, containsNegated, out var addOrRemoveInvocation) ||
+                    !AreInvocationsOnSameInstance(containsInvocation, addOrRemoveInvocation) ||
+                    !AreInvocationArgumentsEqual(containsInvocation, addOrRemoveInvocation))
+                {
+                    return;
+                }
+
+                using var locations = ArrayBuilder<Location>.GetInstance(2);
+                locations.Add(conditional.Syntax.GetLocation());
+                locations.Add(addOrRemoveInvocation.Syntax.Parent!.GetLocation());
+
+                // Build custom format instead of CSharpShortErrorMessageFormat/VisualBasicShortErrorMessageFormat to prevent unhelpful messages for VB.
+                var symbolDisplayFormat = SymbolDisplayFormat.MinimallyQualifiedFormat
+                    .WithParameterOptions(SymbolDisplayParameterOptions.IncludeType)
+                    .WithGenericsOptions(SymbolDisplayGenericsOptions.None)
+                    .WithMemberOptions(SymbolDisplayMemberOptions.IncludeParameters | SymbolDisplayMemberOptions.IncludeContainingType)
+                    .WithKindOptions(SymbolDisplayKindOptions.None);
```

Comment:
Doesn't seem need to create for each conditional, can be moved to static readonly field

Reply:
Changed in [8a3f6ef](https://github.com/dotnet/roslyn-analyzers/pull/6767/commits/8a3f6ef034caf207bb31272a511b58d877985744).